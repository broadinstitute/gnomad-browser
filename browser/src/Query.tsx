import React, { Component } from 'react'

import Delayed from './Delayed'
import StatusMessage from './StatusMessage'

const areVariablesEqual = (variables: any, otherVariables: any) => {
  const keys = Object.keys(variables)
  const otherKeys = Object.keys(otherVariables)
  if (keys.length !== otherKeys.length) {
    return false
  }
  return keys.every((key) => variables[key] === otherVariables[key])
}

const cancelable = (promise: any) => {
  let isCanceled = false
  const wrapper = new Promise((resolve: any, reject: any) => {
    promise.then(
      (value: any) => {
        if (!isCanceled) {
          resolve(value)
        }
      },
      (error: any) => {
        if (!isCanceled) {
          reject(error)
        }
      }
    )
  })

  return {
    cancel: () => {
      isCanceled = true
    },
    promise: wrapper,
  }
}

type BaseQueryState = any

type BaseQueryProps = {
  operationName: string | null
  query: string
  url: string
  variables?: any
  children: (state: BaseQueryState) => JSX.Element
}

export class BaseQuery extends Component<BaseQueryProps, BaseQueryState> {
  static defaultProps = {
    url: '/api/',
    operationName: null,
    variables: {},
  }

  currentRequest: any

  mounted: any

  state = {
    data: null,
    error: null,
    graphQLErrors: null,
    loading: true,
  }

  componentDidMount() {
    this.mounted = true
    this.loadData()
  }

  componentDidUpdate(prevProps: BaseQueryProps) {
    const { query, variables } = this.props
    if (query !== prevProps.query || !areVariablesEqual(variables, prevProps.variables)) {
      this.loadData()
    }
  }

  componentWillUnmount() {
    this.mounted = false
  }

  loadData() {
    const { operationName, query, url, variables } = this.props

    this.setState({
      loading: true,
      error: null,
      graphQLErrors: null,
    })

    if (this.currentRequest) {
      this.currentRequest.cancel()
    }

    this.currentRequest = cancelable(
      fetch(url, {
        body: JSON.stringify({
          operationName,
          query,
          variables,
        }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
      }).then((response) => response.json())
    )
    this.currentRequest.promise.then(
      (response: any) => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: response.data,
          error: null,
          graphQLErrors: response.errors,
          loading: false,
        })
      },
      (error: any) => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: null,
          error,
          graphQLErrors: null,
          loading: false,
        })
      }
    )
  }

  render() {
    const { children } = this.props
    return children(this.state)
  }
}

type OwnQueryProps = {
  children: (...args: any[]) => any
  errorMessage?: string
  loadingMessage?: string
  loadingPlaceholderHeight?: number
  operationName: string | null
  query: string
  success?: (...args: any[]) => any
  url: string
  variables?: any
}

// @ts-expect-error TS(2456) FIXME: Type alias 'QueryProps' circularly references itse... Remove this comment to see the full error message
type QueryProps = OwnQueryProps & typeof Query.defaultProps

// @ts-expect-error TS(7022) FIXME: 'Query' implicitly has type 'any' because it does ... Remove this comment to see the full error message
const Query = ({
  children,
  errorMessage,
  loadingMessage,
  loadingPlaceholderHeight,
  operationName,
  query,
  success,
  url,
  variables,
}: QueryProps) => {
  return (
    <BaseQuery operationName={operationName} query={query} url={url} variables={variables}>
      {({ data, error, graphQLErrors, loading }: any) => {
        if (loading) {
          return (
            <div style={{ height: loadingPlaceholderHeight || 'auto' }}>
              <Delayed>
                <StatusMessage>{loadingMessage}</StatusMessage>
              </Delayed>
            </div>
          )
        }

        if (error) {
          return <StatusMessage>{errorMessage}</StatusMessage>
        }

        if (!data || !success(data)) {
          return (
            <StatusMessage>
              {graphQLErrors && graphQLErrors.length
                ? Array.from(new Set(graphQLErrors.map((e: any) => e.message))).join(', ')
                : errorMessage}
            </StatusMessage>
          )
        }

        return children({ data })
      }}
    </BaseQuery>
  )
}

Query.defaultProps = {
  errorMessage: 'Error',
  loadingMessage: 'Loading',
  loadingPlaceholderHeight: undefined,
  success: () => true,
  url: '/api/',
  variables: {},
  operationName: null,
}

export default Query
