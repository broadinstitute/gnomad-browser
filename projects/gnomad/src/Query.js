import gqlFetch from 'graphql-fetch'
import PropTypes from 'prop-types'
import { Component } from 'react'

const areVariablesEqual = (variables, otherVariables) => {
  const keys = Object.keys(variables)
  const otherKeys = Object.keys(otherVariables)
  if (keys.length !== otherKeys.length) {
    return false
  }
  return keys.every(key => variables[key] === otherVariables[key])
}

const cancelable = promise => {
  let isCanceled = false
  const wrapper = new Promise((resolve, reject) => {
    promise.then(
      value => {
        if (!isCanceled) {
          resolve(value)
        }
      },
      error => {
        if (!isCanceled) {
          reject(error)
        }
      }
    )
  })

  return {
    cancel: () => {
      isCanceled = true
    },
    promise: wrapper,
  }
}

export class Query extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
    variables: PropTypes.object,
  }

  static defaultProps = {
    variables: {},
  }

  state = {
    data: null,
    error: null,
    loading: true,
  }

  componentDidMount() {
    this.mounted = true
    this.loadData()
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.query !== prevProps.query ||
      !areVariablesEqual(this.props.variables, prevProps.variables)
    ) {
      this.loadData()
    }
  }

  componentWillUnmount() {
    this.mounted = false
  }

  loadData() {
    this.setState({
      loading: true,
      error: null,
    })

    if (this.currentRequest) {
      this.currentRequest.cancel()
    }

    this.currentRequest = cancelable(
      gqlFetch(process.env.GNOMAD_API_URL)(this.props.query, this.props.variables)
    )
    this.currentRequest.promise.then(
      response => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: response.data,
          error: null,
          loading: false,
        })
      },
      error => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: null,
          error,
          loading: false,
        })
      }
    )
  }

  render() {
    return this.props.children(this.state)
  }
}
