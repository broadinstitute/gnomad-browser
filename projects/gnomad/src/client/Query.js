import gqlFetch from 'graphql-fetch'
import PropTypes from 'prop-types'
import { Component } from 'react'

const areVariablesEqual = (variables, otherVariables) => {
  const keys = Object.keys(variables)
  const otherKeys = Object.keys(otherVariables)
  if (keys.length !== otherKeys.length) {
    return false
  }
  return keys.every(key => variables[key] === otherVariables[key])
}

const cancelable = promise => {
  let isCanceled = false
  const wrapper = new Promise((resolve, reject) => {
    promise.then(
      value => {
        if (!isCanceled) {
          resolve(value)
        }
      },
      error => {
        if (!isCanceled) {
          reject(error)
        }
      }
    )
  })

  return {
    cancel: () => {
      isCanceled = true
    },
    promise: wrapper,
  }
}

class Query extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
    url: PropTypes.string,
    variables: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  }

  static defaultProps = {
    url: '/api/',
    variables: {},
  }

  state = {
    data: null,
    error: null,
    graphQLErrors: null,
    loading: true,
  }

  componentDidMount() {
    this.mounted = true
    this.loadData()
  }

  componentDidUpdate(prevProps) {
    const { query, variables } = this.props
    if (query !== prevProps.query || !areVariablesEqual(variables, prevProps.variables)) {
      this.loadData()
    }
  }

  componentWillUnmount() {
    this.mounted = false
  }

  loadData() {
    const { query, url, variables } = this.props

    this.setState({
      loading: true,
      error: null,
      graphQLErrors: null,
    })

    if (this.currentRequest) {
      this.currentRequest.cancel()
    }

    this.currentRequest = cancelable(gqlFetch(url)(query, variables))
    this.currentRequest.promise.then(
      response => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: response.data,
          error: null,
          graphQLErrors: response.errors,
          loading: false,
        })
      },
      error => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: null,
          error,
          graphQLErrors: null,
          loading: false,
        })
      }
    )
  }

  render() {
    const { children } = this.props
    return children(this.state)
  }
}

export default Query
