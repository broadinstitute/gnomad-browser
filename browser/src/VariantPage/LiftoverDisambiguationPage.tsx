import React from 'react'
import {
  DatasetId,
  labelForDataset,
  referenceGenome,
  isLiftoverSource,
  baseDatasetForReferenceGenome,
} from '@gnomad/dataset-metadata/metadata'
import { TrackPage, TrackPageSection } from '../TrackPage'
import DocumentTitle from '../DocumentTitle'
import { BaseQuery } from '../Query'
import Delayed from '../Delayed'
import StatusMessage from '../StatusMessage'
import Link from '../Link'
import { Redirect } from 'react-router-dom'

type LiftoverDisambiguationPageProps = {
  fromVariantId: string
  fromDatasetId: DatasetId
  toDatasetId: DatasetId
}

type CorrespondingVariantField = 'liftover' | 'source'

const LiftoverDisambiguationPage = ({
  fromVariantId,
  fromDatasetId,
  toDatasetId,
}: LiftoverDisambiguationPageProps) => {
  const fromLabel = labelForDataset(fromDatasetId)
  const fromReferenceGenome = referenceGenome(fromDatasetId)
  const toReferenceGenome = referenceGenome(toDatasetId)

  const operationName = 'LiftoverDisambiguation'
  const correspondingVariantField: CorrespondingVariantField = isLiftoverSource(fromDatasetId)
    ? 'liftover'
    : 'source'
  const disambiguationQuery = `
query ${operationName}($source_variant_id: String, $liftover_variant_id: String, $reference_genome: ReferenceGenomeId!) {
  liftover(source_variant_id: $source_variant_id, liftover_variant_id: $liftover_variant_id, reference_genome: $reference_genome) {
	${correspondingVariantField} {
	  variant_id
	}
    }
  }
`

  const baseQueryVariables = { reference_genome: referenceGenome(fromDatasetId) }
  const queryVariables =
    fromReferenceGenome === 'GRCh37'
      ? { ...baseQueryVariables, source_variant_id: fromVariantId }
      : { ...baseQueryVariables, liftover_variant_id: fromVariantId }

  return (
    <TrackPage>
      <TrackPageSection>
        <DocumentTitle
          title={`${fromVariantId} liftover | ${fromReferenceGenome} to ${toReferenceGenome}`}
        />
        <>
          Due to liftover, variant {fromVariantId} in dataset {fromLabel} may correspond to a
          different variant or variants in {toReferenceGenome}.
        </>
        <BaseQuery
          operationName="LiftoverDisambiguation"
          query={disambiguationQuery}
          variables={queryVariables}
        >
          {({ data, error, graphQLErrors, loading }: any) => {
            let pageContent = null
            if (loading) {
              pageContent = (
                <Delayed>
                  <StatusMessage>Loading corresponding variants...</StatusMessage>
                </Delayed>
              )
            } else if (error) {
              pageContent = <StatusMessage>Unable to load corresponding variants</StatusMessage>
            } else if (!(data || {}).liftover) {
              pageContent = (
                <StatusMessage>
                  {graphQLErrors && graphQLErrors.length
                    ? Array.from(
                        new Set(
                          graphQLErrors
                            .filter((e: any) => !e.message.includes('ClinVar'))
                            .map((e: any) => e.message)
                        )
                      ).join(', ')
                    : 'Unable to load corresponding variants'}
                </StatusMessage>
              )
            } else if (data.liftover.length === 0) {
              pageContent = <StatusMessage>No corresponding variants found</StatusMessage>
            } else if (data.liftover.length === 1) {
              pageContent = (
                <Redirect
                  to={{
                    pathname: `/variant/${data.liftover[0][correspondingVariantField].variant_id}`,
                    search: `dataset=${baseDatasetForReferenceGenome(toReferenceGenome)}`,
                  }}
                />
              )
            } else {
              pageContent = (
                <>
                  <ul>
                    {data.liftover.map(
                      (
                        correspondingVariant: Record<
                          CorrespondingVariantField,
                          { variant_id: string } | undefined
                        >
                      ) => {
                        const variantId =
                          correspondingVariant[correspondingVariantField]!.variant_id
                        const datasetToLink = baseDatasetForReferenceGenome(toReferenceGenome)
                        return (
                          <li key={variantId}>
                            <Link
                              to={`/variant/${variantId}?dataset=${datasetToLink}`}
                              preserveSelectedDataset={false}
                            >
                              {variantId}
                            </Link>
                          </li>
                        )
                      }
                    )}
                  </ul>
                </>
              )
            }

            return <>{pageContent}</>
          }}
        </BaseQuery>
      </TrackPageSection>
    </TrackPage>
  )
}

export default LiftoverDisambiguationPage
