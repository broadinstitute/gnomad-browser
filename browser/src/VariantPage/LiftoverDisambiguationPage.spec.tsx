import React from 'react'
import { jest, describe, expect, test, beforeEach, afterEach } from '@jest/globals'
import renderer from 'react-test-renderer'
import { mockQueries } from '../../../tests/__helpers__/queries'
import { withDummyRouter } from '../../../tests/__helpers__/router'
import Query, { BaseQuery } from '../Query'
import LiftoverDisambiguationPage from './LiftoverDisambiguationPage'
import { createMemoryHistory } from 'history'

import {
  DatasetId,
  allDatasetIds,
  isLiftoverSource,
  isLiftoverTarget,
} from '@gnomad/dataset-metadata/metadata'

jest.mock('../Query', () => {
  const originalModule = jest.requireActual('../Query')

  return {
    __esModule: true,
    ...(originalModule as object),
    default: jest.fn(),
    BaseQuery: jest.fn(),
  }
})

const {
  resetMockApiCalls,
  resetMockApiResponses,
  simulateApiResponse,
  setMockApiResponses,
  mockApiCalls,
} = mockQueries()

beforeEach(() => {
  Query.mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('Query', query, children, operationName, variables)
    )
  )
  ;(BaseQuery as any).mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('BaseQuery', query, children, operationName, variables)
    )
  )
})

afterEach(() => {
  resetMockApiCalls()
  resetMockApiResponses()
})

const liftoverSourceDatasets = allDatasetIds.filter(isLiftoverSource)
const liftoverTargetDatasets = allDatasetIds.filter(isLiftoverTarget)

describe('LiftoverDisambiguationPage', () => {
  describe.each(liftoverSourceDatasets)(
    'starting from liftover source dataset %s',
    (fromDatasetId: DatasetId) => {
      test('makes the correct query', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({}),
        })

        renderer.create(
          withDummyRouter(
            <LiftoverDisambiguationPage
              fromVariantId="fakevariant"
              fromDatasetId={fromDatasetId}
              toDatasetId="gnomad_r4"
            />
          )
        )

        const queries = mockApiCalls()
        expect(queries.length).toEqual(1)

        const query = queries[0]
        expect(query.operationName).toEqual('LiftoverDisambiguation')
        expect(query.variables.source_variant_id).toEqual('fakevariant')
        expect(query.variables.liftover_variant_id).toEqual(undefined)
        expect(query.variables.reference_genome).toEqual('GRCh37')
      })
      test('has appropriate message if no corresponding variant found', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({ liftover: [] }),
        })

        const tree = renderer.create(
          withDummyRouter(
            <LiftoverDisambiguationPage
              fromVariantId="fakevariant"
              fromDatasetId={fromDatasetId}
              toDatasetId="gnomad_r4"
            />
          )
        )
        expect(tree).toMatchSnapshot()
      })

      test('redirects if one corresponding variant found', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({
            liftover: [{ liftover: { variant_id: 'source1' } }],
          }),
        })

        const history = createMemoryHistory()

        const router = withDummyRouter(
          <LiftoverDisambiguationPage
            fromVariantId="fakevariant"
            fromDatasetId={fromDatasetId}
            toDatasetId="gnomad_r4"
          />,
          history
        )
        renderer.create(router)
        const { location } = history
        expect(location.pathname).toEqual('/variant/source1')
        expect(location.search).toEqual('?dataset=gnomad_r4')
      })

      test('renders links to each variant if multiple corresponding variants found', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({
            liftover: [
              { liftover: { variant_id: 'source1' } },
              { liftover: { variant_id: 'source2' } },
            ],
          }),
        })

        const tree = renderer.create(
          withDummyRouter(
            <LiftoverDisambiguationPage
              fromVariantId="fakevariant"
              fromDatasetId={fromDatasetId}
              toDatasetId="gnomad_r4"
            />
          )
        )

        expect(tree).toMatchSnapshot()
      })
    }
  )

  describe.each(liftoverTargetDatasets)(
    'starting from liftover target dataset %s',
    (fromDatasetId: DatasetId) => {
      test('makes the correct query', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({}),
        })

        renderer.create(
          withDummyRouter(
            <LiftoverDisambiguationPage
              fromVariantId="fakevariant"
              fromDatasetId={fromDatasetId}
              toDatasetId="gnomad_r2_1"
            />
          )
        )

        const queries = mockApiCalls()
        expect(queries.length).toEqual(1)

        const query = queries[0]
        expect(query.operationName).toEqual('LiftoverDisambiguation')
        expect(query.variables.source_variant_id).toEqual(undefined)
        expect(query.variables.liftover_variant_id).toEqual('fakevariant')
        expect(query.variables.reference_genome).toEqual('GRCh38')
      })

      test('has appropriate message if no corresponding variant found', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({ liftover: [] }),
        })

        const tree = renderer.create(
          withDummyRouter(
            <LiftoverDisambiguationPage
              fromVariantId="fakevariant"
              fromDatasetId={fromDatasetId}
              toDatasetId="gnomad_r2_1"
            />
          )
        )
        expect(tree).toMatchSnapshot()
      })

      test('redirects if one corresponding variant found', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({
            liftover: [{ source: { variant_id: 'source1' } }],
          }),
        })

        const history = createMemoryHistory()

        const router = withDummyRouter(
          <LiftoverDisambiguationPage
            fromVariantId="fakevariant"
            fromDatasetId={fromDatasetId}
            toDatasetId="gnomad_r2_1"
          />,
          history
        )
        renderer.create(router)
        const { location } = history
        expect(location.pathname).toEqual('/variant/source1')
        expect(location.search).toEqual('?dataset=gnomad_r2_1')
      })

      test('renders links to each variant if multiple corresponding variants found', () => {
        setMockApiResponses({
          LiftoverDisambiguation: () => ({
            liftover: [
              { source: { variant_id: 'source1' } },
              { source: { variant_id: 'source2' } },
            ],
          }),
        })

        const tree = renderer.create(
          withDummyRouter(
            <LiftoverDisambiguationPage
              fromVariantId="fakevariant"
              fromDatasetId={fromDatasetId}
              toDatasetId="gnomad_r2_1"
            />
          )
        )

        expect(tree).toMatchSnapshot()
      })
    }
  )
})
