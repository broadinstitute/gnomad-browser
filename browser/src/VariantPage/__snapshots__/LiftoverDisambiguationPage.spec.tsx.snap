// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1 has appropriate message if no corresponding variant found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1
     may correspond to a different variant or variants in 
    GRCh38
    .
    <div
      className="StatusMessage-sc-neg7mh-0 hJFlwG"
    >
      No corresponding variants found
    </div>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1 renders links to each variant if multiple corresponding variants found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1
     may correspond to a different variant or variants in 
    GRCh38
    .
    <ul>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source1?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source1
        </a>
      </li>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source2?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source2
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_controls has appropriate message if no corresponding variant found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (controls)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <div
      className="StatusMessage-sc-neg7mh-0 hJFlwG"
    >
      No corresponding variants found
    </div>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_controls renders links to each variant if multiple corresponding variants found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (controls)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <ul>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source1?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source1
        </a>
      </li>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source2?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source2
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_non_cancer has appropriate message if no corresponding variant found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (non-cancer)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <div
      className="StatusMessage-sc-neg7mh-0 hJFlwG"
    >
      No corresponding variants found
    </div>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_non_cancer renders links to each variant if multiple corresponding variants found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (non-cancer)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <ul>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source1?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source1
        </a>
      </li>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source2?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source2
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_non_neuro has appropriate message if no corresponding variant found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (non-neuro)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <div
      className="StatusMessage-sc-neg7mh-0 hJFlwG"
    >
      No corresponding variants found
    </div>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_non_neuro renders links to each variant if multiple corresponding variants found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (non-neuro)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <ul>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source1?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source1
        </a>
      </li>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source2?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source2
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_non_topmed has appropriate message if no corresponding variant found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (non-TOPMed)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <div
      className="StatusMessage-sc-neg7mh-0 hJFlwG"
    >
      No corresponding variants found
    </div>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover source dataset gnomad_r2_1_non_topmed renders links to each variant if multiple corresponding variants found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v2.1.1 (non-TOPMed)
     may correspond to a different variant or variants in 
    GRCh38
    .
    <ul>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source1?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source1
        </a>
      </li>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source2?dataset=gnomad_r4"
          onClick={[Function]}
        >
          source2
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover target dataset gnomad_r4 has appropriate message if no corresponding variant found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v4.0.0
     may correspond to a different variant or variants in 
    GRCh37
    .
    <div
      className="StatusMessage-sc-neg7mh-0 hJFlwG"
    >
      No corresponding variants found
    </div>
  </div>
</div>
`;

exports[`LiftoverDisambiguationPage starting from liftover target dataset gnomad_r4 renders links to each variant if multiple corresponding variants found 1`] = `
<div
  className="Page-w7h773-0 TrackPage-sc-1xq3qi7-0 dWdnsu"
>
  <div
    className="TrackPage__TrackPageSection-sc-1xq3qi7-1 fQVUsz"
  >
    Due to liftover, variant 
    fakevariant
     in dataset 
    gnomAD v4.0.0
     may correspond to a different variant or variants in 
    GRCh37
    .
    <ul>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source1?dataset=gnomad_r2_1"
          onClick={[Function]}
        >
          source1
        </a>
      </li>
      <li>
        <a
          className="Link-sc-14lgydv-0-Link jBvaYQ"
          href="/variant/source2?dataset=gnomad_r2_1"
          onClick={[Function]}
        >
          source2
        </a>
      </li>
    </ul>
  </div>
</div>
`;
